distributed/formats/tests/test_csv.py

```diff
@@ -0,0 +1,134 @@
+from distributed.compatibility import gzip_compress
+from distributed.utils_test import gen_cluster
+from distributed.formats.csv import read_csv, bytes_read_csv
+from distributed.executor import Future
+
+
+import pandas.util.testing as tm
+import pandas as pd
+from tornado import gen
+import pytest
+from io import BytesIO
+from toolz import partition_all
+
+dd = pytest.importorskip('dask.dataframe')
+
+
+files = {'2014-01-01.csv': (b'name,amount,id\n'
+                            b'Alice,100,1\n'
+                            b'Bob,200,2\n'
+                            b'Charlie,300,3\n'),
+         '2014-01-02.csv': (b'name,amount,id\n'),
+         '2014-01-03.csv': (b'name,amount,id\n'
+                            b'Dennis,400,4\n'
+                            b'Edith,500,5\n'
+                            b'Frank,600,6\n')}
+
+
+header = files['2014-01-01.csv'].split(b'\n')[0] + b'\n'
+
+expected = pd.concat([pd.read_csv(BytesIO(files[k])) for k in sorted(files)])
+
+
+def test_bytes_read_csv():
+    b = files['2014-01-01.csv']
+    df = bytes_read_csv(b, '', {})
+    assert list(df.columns) == ['name', 'amount', 'id']
+    assert len(df) == 3
+    assert df.id.sum() == 1 + 2 + 3
+
+
+def test_bytes_read_csv_kwargs():
+    b = files['2014-01-01.csv']
+    df = bytes_read_csv(b, '', {'usecols': ['name', 'id']})
+    assert list(df.columns) == ['name', 'id']
+
+
+def test_bytes_read_csv_with_header():
+    b = files['2014-01-01.csv']
+    header, b = b.split(b'\n', 1)
+    df = bytes_read_csv(b, header, {})
+    assert list(df.columns) == ['name', 'amount', 'id']
+    assert len(df) == 3
+    assert df.id.sum() == 1 + 2 + 3
+
+
+
+@gen_cluster(executor=True)
+def test_read_csv(e, s, a, b):
+    bytes = [files[k] for k in sorted(files)]
+    gzbytes = [gzip_compress(b) for b in bytes]
+    kwargs = {}
+    head = bytes_read_csv(files['2014-01-01.csv'], '', {})
```

Comment:
The intention is that the user never sees this code.  This is for use by other internal modules, like `s3` and `hdfs`, who will have their own, simpler, `read_csv` functions.